src/platform/FailSafeContext.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides the implementation of the FailSafeContext object.
+ */
+
+#include <platform/FailSafeContext.h>
+
+#include <platform/ConfigurationManager.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+void FailSafeContext::HandleArmFailSafe(System::Layer * layer, void * aAppState)
+{
+    FailSafeContext * context = reinterpret_cast<FailSafeContext *>(aAppState);
+    context->CommissioningFailedTimerComplete();
+}
+
+void FailSafeContext::CommissioningFailedTimerComplete()
+{
+    // TODO: If the fail-safe timer expires before the CommissioningComplete command is
+    // successfully invoked, conduct clean-up steps.
+
+    ChipDeviceEvent event;
+    event.Type                         = DeviceEventType::kCommissioningComplete;
+    event.CommissioningComplete.status = CHIP_ERROR_TIMEOUT;
+    CHIP_ERROR status                  = PlatformMgr().PostEvent(&event);
+
+    mFailSafeArmed            = false;
+    mNocCommandHasBeenInvoked = false;
+
+    if (status != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "Failed to post commissioning complete: %" CHIP_ERROR_FORMAT, status.Format());
+    }
+}
+
+CHIP_ERROR FailSafeContext::ArmFailSafe(FabricIndex accessingFabric, System::Clock::Timeout expiryLength)
+{
+    VerifyOrReturnError(MatchedFabricIndex(accessingFabric), CHIP_ERROR_ACCESS_DENIED);
```

Comment:
I don't understand this check.  Why are we checking whatever stale value  mFabricIndex has here?  It might not have anything to do with who is now arming the failsafe.

We _could_ probably check that either `!mFailSafeArmed` or we match the accessing fabric index....

Reply:
We are checking the accessing fabric matches the fail-safe context’s Fabric Index, no?
VerifyOrReturnError(MatchedFabricIndex(accessingFabric), CHIP_ERROR_ACCESS_DENIED);

Reply:
But the fail-safe context's fabric index is only meaningful _when the fail-safe is armed_, no?

Reply:
Move this logic check out and do it in the caller side